magento/pwa-studio
# [PWA-2235] Apollo cache eviction with multiple store view support

## Description

[PWA-962](https://github.com/magento/pwa-studio/pull/3397) changed the cache clearing mechanism for sensitive data to use v3 Apollo's built in cache eviction. The problem is that this eviction will only clear for the current persistor. This regression was caught during the release process and since the previous mechanism was working correctly and the change didn't bring any new functionality, we reverted it for the release.

The goal of this task is to re-apply the v3 cache eviction, and find a workaround for clearing the inactive store view's persisted cache.

---

Refactoring cache clearing of customer and cart data to use Apollo v3's `evict` and garbage cleaning.

## Related Issue







Closes [PWA-2235](https://jira.corp.magento.com/browse/PWA-2235)

## Acceptance

- Utilities for manual cache deletion are deprecated - deleteCacheEntry
- These calls are replaced by cache.evict(), targeting specific entries that need purged
- After checking out, all references to cart are no longer present in InMemory or persisted cache in localStorage  **for all available stores**
- After logging out, all references to cart and customer data are no longer present in InMemory or persisted cache in localStorage **for all available stores**
- After creating an account, all references to cart and customer data are no longer present in InMemory or persisted cache in localStorage **in the other stores**
- After signing in, all references to cart and customer data are no longer present in InMemory or persisted cache in localStorage **in the other stores**
- Order History data should be removed from cache when signing out.

### Verification Stakeholders



## Verification Steps






#### Test scenario(s) for direct fix/feature

Test 1
- Sign-in as a customer
- Verify localStorage to see Apollo cache entries for customer data
- Add a new customer address (through checkout or through account)
- Add items to cart
- Verify cart data in localStorage
- Place order
- Entries for cart should be cleared (Cart will remain but with new ID)
- Sign out
- Entries for customer data should no longer be in cache

Test 2
- Sign-in as a customer
- Verify localStorage to see Apollo cache entries for customer data
- Add a new customer address (through checkout or through account)
- Add items to cart
- Verify cart data in localStorage
- Checkout a product, fill out all details on the checkout page but do not complete the order yet
- Switch to the `french` store
- Sign out
- Entries for cart and customer data should no longer be in cache

Test 3
- Navigate as guest
- Add items to cart
- Switch to the `french` store
- Verify cart data in localStorage in both stores
- Create a new account
- Entries for cart and customer data should no longer be in cache of the `default` store (They will be regenerated when you switch to that store)

Test 4
- Login
- Checkout an item and place an order on the default store
- Go to the order history page
- Validate in the Apollo caches that you have OrderItem data entries
- Switch to another store
- Logout
- Validate in the Apollo caches that you don't have OrderItem data entries remaining

## Breaking Changes (if any)

Removed `deleteCacheEntry` file/function

## Checklist
